Adding alpha

DavidRGriswold opened the pull request.

Commit:
added CSS for alpha panel (not implemented in this commit)

Commit:
refactored, added allowAlpha command with alpha slider and backgrounds, got all features implemented as before otherwise.

Commit:
forgot to commit js

Commit:
fixed css for palette borders

Commit:
playing with CSS for testing and such

DavidRGriswold committed:
switching to master.

DavidRGriswold committed:
fully operaional and tested!

DavidRGriswold committed:
Final commit of alpha channel, fully implemented.

An alpha channel is now supported!
-- use the "allowAlpha: true" option to turn on an alpha bar if palette is square
-- by default, returns an 8-digit hex code suitable for R (or very modern css). Can also return names.
-- alpha channels are also allowed in allowedCols - use 4- or 8-digit hex, hsla, or rgba() notation.

Other color formats permitted as inputs
-- in showColour = text or showColour = both modes, you can input colors as names, or in hex (3, 4, 6, 8), rgb(a), or hsl(a) modes

Other Changes:
-- allowedCols should now be entered as an array of strings (it will parse a string if entered as one)
-- "palette=limited" mode is now slightly wider
-- IE8 and IE9 support are now more CSS-based

DavidRGriswold committed:
Final commit of alpha channel, fully implemented.

An alpha channel is now supported!
-- use the "allowAlpha: true" option to turn on an alpha bar if palette is square
-- by default, returns an 8-digit hex code suitable for R (or very modern css). Can also return names.
-- alpha channels are also allowed in allowedCols - use 4- or 8-digit hex, hsla, or rgba() notation.

Other color formats permitted as inputs
-- in showColour = text or showColour = both modes, you can input colors as names, or in hex (3, 4, 6, 8), rgb(a), or hsl(a) modes

Other Changes:
-- allowedCols should now be entered as an array of strings (it will parse a string if entered as one)
-- "palette=limited" mode is now slightly wider
-- IE8 and IE9 support are now more CSS-based

DavidRGriswold committed:
Merge branch 'master' of github.com:DavidRGriswold/jquery-colourpicker

DavidRGriswold committed:
Final commit of alpha channel, fully implemented, now with readme and minified files!

An alpha channel is now supported!
-- use the "allowAlpha: true" option to turn on an alpha bar if palette is square
-- by default, returns an 8-digit hex code suitable for R (or very modern css). Can also return names.
-- alpha channels are also allowed in allowedCols - use 4- or 8-digit hex, hsla, or rgba() notation.

Other color formats permitted as inputs
-- in showColour = text or showColour = both modes, you can input colors as names, or in hex (3, 4, 6, 8), rgb(a), or hsl(a) modes

Other Changes:
-- allowedCols should now be entered as an array of strings (it will parse a string if entered as one)
-- "palette=limited" mode is now slightly wider
-- IE8 and IE9 support are now more CSS-based

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
+  padding-left: 0px;
+  padding-right: 0px;
+  padding-top: 1px;
+  padding-bottom: 4px;
```
If all 4 are defined, could you merge them into one statement `padding: 1px 0 4px 0;`

DavidRGriswold replied:
sure, that makes sense.

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
 }
-.colourpicker .cp-list-row {
+
+/*.colourpicker .cp-list-row {
```
instead of commenting out, just delete it. That's what git is for :) Codebase will become messy quick if changes are commented instead of removed

DavidRGriswold replied:
got it.

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
+/*.colourpicker .cp-list-row {
   margin-top: 3px;
-}
+*/}
```
shouldn't the comment end AFTER the closing brace?

DavidRGriswold replied:
um yes. Good catch.

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
+  height: 100%;
+  width: 100%;
+  display:block;
+  /*border: 1px solid #DDDDDD;*/
```
remove commented lines

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
+.colourpicker .cp-list-col-outer {
   width: 20px;
   height: 20px;
+  margin-top: 3px;
```
if margin left is 4, then margin top should also be 4. It looks a bit off when the vertical and horizontal spacing isn't uniform. Either both at 3 or both at 4.

DavidRGriswold replied:
ok

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
-.colourpicker .cp-list-col:first-child {
-  margin-left: 0;
-}
+/*.colourpicker .cp-list-col-outer:first-child {
```
remove

daattali commented on css/colourpicker.css:
```diff
@@ -36,36 +42,59 @@
 .colourpicker .cp-list-col {
+  height: 100%;
+  width: 100%;
+  display:block;
```
whitespace between property name and value (I know it sounds anal, but I'm very big on readability over the long run)

DavidRGriswold replied:
=)

daattali commented on css/colourpicker.css:
```diff
@@ -79,12 +108,19 @@
   width: 100%;
   height: 100%;
+  background: linear-gradient(to right,#FFF,rgba(255,255,255,0));
+  filter:progid:DXImageTransform.Microsoft.gradient(gradientType=1,startColorstr=#FFFFFFFF,endColorstr=#00000000);
```
whitespace after filter

daattali commented on css/colourpicker.css:
```diff
@@ -79,12 +108,19 @@
+}
+
+.colourpicker .colourpicker-grid-inner-2 {
+  width:100%;
```
whitespace (you can run the CSS file through a CSS linter/beautifier instead of manually adding the spaces. I won't add this comment anymore)

DavidRGriswold replied:
suggestions for a good beautifier? Command-line preferred but I suppose I can use a web tool if I must. Maybe there's a sublime add in...

daattali commented on css/colourpicker.css:
```diff
@@ -130,17 +166,49 @@
+.colourpicker-alpha-inner-slider {
+  display:none;
+  position:absolute;
+  top:0;left:0;bottom:0;right:0;
```
4 separate lines

DavidRGriswold replied:
stole that from minicolour. will fix.

daattali commented on index.html:
```diff
@@ -29,20 +29,47 @@
+    {palette: "limited", returnName: true}<br>
     Value: <span></span>
     <input type="text" id="col4" class="form-control my-cp">
+  </div><div>
```
start the div on new line (unless there's a reason for having it on same line?)

daattali commented on js/colourpicker.js:
```diff
@@ -38,21 +38,22 @@
-          "#00868B #008B45 #458B00 #008B00 #00FF00 #7FFF00 #54FF9F #00FF7F " +
-          "#7FFFD4 #8B4500 #8B0000 #FF0000 #FF6A6A #FF7F00 #FFFF00 #FFF68F " +
-          "#F4A460 #551A8B #8B008B #8B0A50 #9400D3 #FF00FF #FF1493 #E066FF",
+          ["#333333FF", "#4D4D4DFF", "#666666FF", "#7F7F7FFF", "#999999FF", "#B3B3B3FF", "#E5E5E5FF",  
```
any reason the first color #000000 was removed?

DavidRGriswold replied:
must have been copy and paste error. Explains why there were only 39! I'll add it back int.

daattali replied:
testing something: #000000

daattali replied:
woah this is a strange bug in github. If you type `#000000` into a comment, it only shows one zero!

daattali replied:
WOAH and this is even cooler, looks like putting a hex code inside a code block makes github show the resulting colour beside it!

DavidRGriswold replied:
very cool!

daattali commented on js/colourpicker.js:
```diff
@@ -38,21 +38,22 @@
-        colsMap:
-          {"#FFFFFF":"white","#F0F8FF":"aliceblue","#FAEBD7":"antiquewhite","#FFEFDB":"antiquewhite1","#EEDFCC":"antiquewhite2","#CDC0B0":"antiquewhite3","#8B8378":"antiquewhite4","#7FFFD4":"aquamarine","#76EEC6":"aquamarine2","#66CDAA":"aquamarine3","#458B74":"aquamarine4","#F0FFFF":"azure","#E0EEEE":"azure2","#C1CDCD":"azure3","#838B8B":"azure4","#F5F5DC":"beige","#FFE4C4":"bisque","#EED5B7":"b ...
-
+        //colsMap: No longer needed, hard-coded in below instead
```
remove commented code

daattali commented on js/colourpicker.js:
```diff
@@ -38,21 +38,22 @@
-
+        //colsMap: No longer needed, hard-coded in below instead
+        //  {"#FFFFFF":"white","#F0F8FF":"aliceblue","#FAEBD7":"antiquewhite","#FFEFDB":"antiquewhite1","#EEDFCC":"antiquewhite2","#CDC0B0":"antiquewhite3","#8B8378":"antiquewhite4","#7FFFD4":"aquamarine","#76EEC6":"aquamarine2","#66CDAA":"aquamarine3","#458B74":"aquamarine4","#F0FFFF":"azure","#E0EEEE":"azure2","#C1CDCD":"azure3","#838B8B":"azure4","#F5F5DC":"beige","#FFE4C4":"bisque","#EED5B7": ...
+        //added by David Griswold
```
add space between // and comment (for all the rest of the comments too)

daattali commented on js/colourpicker.js:
```diff
@@ -153,16 +143,22 @@
-        colourpicker.addClass('input-group');
+      settings = $.extend(false, {}, defaults, settings);
+      //for backwards compatibility
+      if ($.type(settings.allowedCols) === "string") {
```
This is generally a great idea, it's good that you kept backwards compatibility in mind. But in this very specific case I don't think we need to do that because this library is literally only used for the shiny package, which will know to send in the correct format when we use this updated version. I highly doubt anybody else is using this library. If you think someone might be using it and would be affected, then keep this check,

DavidRGriswold replied:
I prefer to keep it. You never know, right? I'm already thinking I might use this in my own javascript projects, so maybe somebody out there secretly did the same thing, unlikely though that is since they'd probably be using mini-color. Still, it's a pretty minor thing.

daattali commented on js/colourpicker.js:
```diff
@@ -153,16 +143,22 @@
+        settings.allowedCols = settings.allowedCols.split(" ");
       }
-
+	    // Palette type, alpha or not, make it an input group.
```
why is it always an input-group? I wanted it to only be an input group when there is actually another UI component embedded inside, so the UI needed to look right, but when there is only one input I don't see the need?

DavidRGriswold replied:
This was the easiest way I could come up with to get a background image of the checkered box. I could add another container class instead, but since this one was already there I decided to leverage it and give input-group the checkered background. However, I see it caused some UI issues further down so I'll reconsider whether this makes sense.

daattali commented on js/colourpicker.js:
```diff
@@ -190,16 +192,30 @@
+          coloursHtml += '<span class="cp-list-col-outer"><span class="cp-list-col" data-cp-col="' + hexstring +'" ' +
+                            'style="' + backgroundstring + '" + ></span></span>';
+        });
+        // coloursHtml += '</div>';
```
remove

daattali commented on js/colourpicker.js:
```diff
@@ -369,129 +386,209 @@
 
       // Helpful references
-      colourpicker = input.parent(),
+      colourpicker = input.parent();
```
did this become a semicolon on purpose?

DavidRGriswold replied:
I was very confused by the long list of variables at some point and made a bunch of commas into semi-colons, but then I figured out what was going on and changed them back. I'll fix this.

daattali commented on js/colourpicker.js:
```diff
@@ -369,129 +386,209 @@
 
       // Panel objects
       grid = colourpicker.find('.colourpicker-grid'),
+      nnergrid = colourpicker.find('.colourpicker-grid-inner'),
```
typo? Is this variable even used anywhere? something strange is happening because I do see references to `innergrid` but that variable is not defined here, yet it works. So many it's not even needed here, and it's defined elsewhere?

DavidRGriswold replied:
perhaps! I'll look into it. Certainly it should be "innergrid"

DavidRGriswold replied:
though honestly, I should probably go back and change the container class situation back to the way you had it originally: since we nixed the checkered background on the grid itself there is no longer any need for two inner containers on the grid.

daattali commented on js/colourpicker.js:
```diff
@@ -369,129 +386,209 @@
-        input.val(hex.toUpperCase());
+        innergrid.css('backgroundColor', hsb2hex({ h: hue, s: 100, b: 100, a : 1}));
+
+        if (alpha !== false) {
```
good type safe checking :+1:

daattali commented on js/colourpicker.js:
```diff
@@ -544,52 +655,375 @@
-      }
-      return hex;
+    // returns an rgb color limited to reasonable values
+    // OR returns false if the 
```
OR return false is the .....?

daattali commented on js/colourpicker.js:
```diff
@@ -544,52 +655,375 @@
+    
+    //giant list of R color names in hex8 format.
+    colsToNames = {"#00000000":"transparent", "#FFFFFFFF":"white","#F0F8FFFF":"aliceblue","#FAEBD7FF":"antiquewhite","#FFEFDBFF":"antiquewhite1","#EEDFCCFF":"antiquewhite2","#CDC0B0FF":"antiquewhite3","#8B8378FF":"antiquewhite4","#7FFFD4FF":"aquamarine","#76EEC6FF":"aquamarine2","#66CDAAFF":"aquamarine3","#458B74FF":"aquamarine4","#F0FFFFFF":"azure","#E0EEEEFF":"azure2","#C1CDCDFF":"azure3","#838 ...
+    function flip(o) {
```
one-line description of what this function does? (Though I get it from looking at the use case)

DavidRGriswold replied:
oh, that was there then got lost in moving functions around. Will add back.

daattali commented on js/colourpicker.js:
```diff
@@ -544,52 +655,375 @@
+    }
+
+        // regex matchers for str2rgb below
+    var matchers = (function() {
```
nice!

DavidRGriswold replied:
I took these from TinyColor. I should probably credit that in the code.

daattali replied:
Ah , yeah those matchers are hardcore, was very impressed haha

daattali commented on js/colourpicker.js:
```diff
@@ -608,101 +1042,117 @@
+    function hsl2hsb(hsl) {
+      return rgb2hsb(hsl2rgb);
+    }
+    function hsl2name(hsl) {
```
Instead of implementing a conversion between every pair of formats, could we just choose one "master" format to convert to and do all others based on that? For example, convert everything to HEX, and then based on the desired return format convert that HEX into rgb\hsl\etc. Does that make sense? Maybe I'm missing something, but I think it'd simplify the code and reduce potential places for error

DavidRGriswold replied:
Most of this code came from TinyColor (again, I should credit that, oops, my credit must have gotten lost). I did add some sanity checks and such. If you look at the functions carefully you'l l see that most of them basically DO do that - they convert everything internally back and forth to rgb. And in practice, all of my calls to it bascially do that too, since the internal colour is stored as rgb. I'll try to remove the unnecessary ones.

daattali reviewed (commented):
Wow, you really did rewrite the JS! I have to admit I don't fully understand all the JS changes, but I have to assume they're correct because I can see it works :) I did have quite a few comments, mostly style, nothing major.

daattali commented:
Some small/tiny UI issues that bug me even though they're very small: (if any of them are unclear or you don't understand, please let me know and I'll add an image)

- in the alpha channel, after clicking on it to move the bar or if you hover over the bar, the cursor turns to default. I think the `cursor: row-resize;` needs to be placed on the parent div `colourpicker-alpha-slider` rather than the `colourpicker-alpha-inner-slider`

- Because the inputs are set to be an input-group always, the right corners (top right and bottom right) are sharp corners instead of rounded, which is inconsistent with the left side and also inconsistent with other bootstrap/shiny inputs

- There's a bit of gray at the top left corner of the input (probably from the checkered background?)

![image](https://user-images.githubusercontent.com/952340/27000488-7da17d1a-4d81-11e7-9394-076cf710751c.png)

- Similar issue: the checkered background image is bleeding a tiny bit from the bottom of the input field. See here, look at the very bottom

![image](https://user-images.githubusercontent.com/952340/27000453-7d08efec-4d80-11e7-9f49-248c23bc4e10.png)

- In palette=square without alpha, the whitespace between the hue bar and the right end of the box is a bit too wide (notice how there is one white pixel strip along the top, bottom, left, and between the hue bar and the square, but to the right of the bar there's 2 or 3 pixels)

![image](https://user-images.githubusercontent.com/952340/27000510-13863be0-4d82-11e7-85d0-7624a70353af.png)

- There seems to be a gray vertical line to the left and to the right of the alpha bar

![image](https://user-images.githubusercontent.com/952340/27000516-402ae696-4d82-11e7-874e-3666721373da.png)


I know all of these are very small UI issues but combined together it takes away a bit from the clean polished feeling

DavidRGriswold commented:
On your UI comments:

1) I will try the row-resize change and see if that fixes it.

2) I did not notice the input-group bleed over. That is interesting, and something I will have to play with. As mentioned earlier, I could probably add a new container class to hold the checkered background instead; maybe I need to do that. I'll twiddle - I'm learning a lot of CSS from this!

3) Oh, yes, I see that extra white space in the square without alpha. Will fix.

4) The gray bars were actually on purpose. I'm not totally wild about it, but I found that without the gray bars it looked like the alpha bar bleeds into the area around it when the color is white. See below. If you like it better as shown below, though, I can easily go back to it.

![image](https://user-images.githubusercontent.com/20600047/27003822-9870a426-4dc3-11e7-852c-6cde58000a92.png)

5) I might make that gray checker image just a tiny bit wider so it tiles more perfectly on the alpha slider. Those tiny half-squares are kind of annoying!

daattali commented:
I don't have a preference for a css beautifier. I usually just find the
first one on Google, just make sure indentation is set to 2 spaces and open
brace is not on its own line
On Jun 10, 2017 10:43 AM, "DavidRGriswold" <notifications@github.com> wrote:

> *@DavidRGriswold* commented on this pull request.
> ------------------------------
>
> In css/colourpicker.css
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121258736>
> :
>
> >  }
>  .colourpicker .colourpicker-grid-inner {
>    width: 100%;
>    height: 100%;
> +  background: linear-gradient(to right,#FFF,rgba(255,255,255,0));
> +  filter:progid:DXImageTransform.Microsoft.gradient(gradientType=1,startColorstr=#FFFFFFFF,endColorstr=#00000000);
> +}
> +
> +.colourpicker .colourpicker-grid-inner-2 {
> +  width:100%;
>
> suggestions for a good beautifier? Command-line preferred but I suppose I
> can use a web tool if I must. Maybe there's a sublime add in...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121258736>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6IFNn9AKSavKutqbWqzyubln-9tdWVks5sCqujgaJpZM4N1sCm>
> .
>

daattali commented:
Sure, keep it
On Jun 10, 2017 10:45 AM, "DavidRGriswold" <notifications@github.com> wrote:

> *@DavidRGriswold* commented on this pull request.
> ------------------------------
>
> In js/colourpicker.js
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121258796>
> :
>
> > @@ -153,16 +143,22 @@
>        if( input.data('colourpicker-initialized') ) return;
>
>        // Handle settings
> -      settings = $.extend(true, {}, defaults, settings);
> -
> -	    // Palette type
> -	    colourpicker.addClass('palette-' + settings.palette);
> -	
> -	    // If adding a transparent checkbox, make the parent an input group
> -      if( settings.allowTransparent ) {
> -        colourpicker.addClass('input-group');
> +      settings = $.extend(false, {}, defaults, settings);
> +      //for backwards compatibility
> +      if ($.type(settings.allowedCols) === "string") {
>
> I prefer to keep it. You never know, right? I'm already thinking I might
> use this in my own javascript projects, so maybe somebody out there
> secretly did the same thing, unlikely though that is since they'd probably
> be using mini-color. Still, it's a pretty minor thing.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121258796>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6IFLSQ82QFU3_YErzZb1kvKB9zYTrTks5sCqwdgaJpZM4N1sCm>
> .
>

DavidRGriswold commented:
Perl was my first programming language so I'm pretty great with regex but
why make what I can take?

On Jun 10, 2017 12:25, "Dean Attali" <notifications@github.com> wrote:

> *@daattali* commented on this pull request.
> ------------------------------
>
> In js/colourpicker.js
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121262054>
> :
>
> >
>      // Keeps value within min and max
>      function keepWithin(value, min, max) {
>        if( value < min ) value = min;
>        if( value > max ) value = max;
>        return value;
>      }
> +    //roundTo function
> +
> +    function roundTo(num,dec) {
> +      return Math.round(num *  Math.pow(10,dec)) / (Math.pow(10,dec));
> +    }
> +
> +        // regex matchers for str2rgb below
> +    var matchers = (function() {
>
> Ah , yeah those matchers are hardcore, was very impressed haha
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#discussion_r121262054>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU79XqrokDo7fM6XLkj_eFVG7LFjIBks5sCtF0gaJpZM4N1sCm>
> .
>

daattali commented:
Regarding the UI stuff: I do think adding a container for the checkered background might solve some of these. 

Alpha channel gray bars: I actually much prefer vit without the bars, the pic you posted looks good!

DavidRGriswold commented:
Okee doke. An easy fix. I will play with the container in a few.

On Jun 10, 2017 12:50, "Dean Attali" <notifications@github.com> wrote:

> Regarding the UI stuff: I do think adding a container for the checkered
> background might solve some of these.
>
> Alpha channel gray bars: I actually much prefer vit without the bars, the
> pic you posted looks good!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#issuecomment-307580301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU7zISYgAcA5yAfAGLydYTfL7Ft_TKks5sCtdqgaJpZM4N1sCm>
> .
>

Commit:
Alpha addition, all fixed. Javascript and CSS cleaned up, UI changes made.
Everything should be good to go, I hope!

Commit:
Noticed a tiny CSS typo as soon as I committed. Trying again!

Commit:
I swear, one of these times it will really be done.

DavidRGriswold committed:
tiny bug in destroy caused big bug in R update. fixed now!

daattali commented:
@DavidRGriswold is this ready for review or still a WIP?

daattali commented:
I just tried playing with it on http://testing.mtbos.org/colour/ and it looks perfect. These changes from today really do take it from 95% to 100% , looks clean and professional

DavidRGriswold commented:
It is ready. I have the coulorpicker R package working too when this goes
through. Check out http://testing.mtbos.org/colour/index.min.html for an
index that uses the minified version to make sure it works too if you want.
I can do one last commit that removes that file if you want.

On Jun 11, 2017 00:58, "Dean Attali" <notifications@github.com> wrote:

> I just tried playing with it on http://testing.mtbos.org/colour/ and it
> looks perfect. These changes from today really do take it from 95% to 100%
> , looks clean and professional
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#issuecomment-307608731>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU7y9ik6CgVEp7ufWXv1pwsi0oUxypks5sC4H4gaJpZM4N1sCm>
> .
>

DavidRGriswold committed:
fixed a width bug that showed up in R testing.

daattali commented:
Yes please do remove that file.

In the future, when you make a non trivial commit to css/javascript, I find it's a good idea to minify them in a separate commit afterwards. Because if you have a large commit and you have both the regular and the min versions in there, it's impossible to know that the min version is actually minified after the last change you made -- maybe you minified and then remembered one more tiny change but forgot to minify again? So to ensure and prove that that's not happening, I prefer to commit, and then minify. No need to remove it this time, just letting you know for future!

daattali committed:
whitespace correction

daattali commented on js/colourpicker.js:
```diff
@@ -668,160 +670,163 @@
     
-    //giant list of R color names in hex8 format.
-    colsToNames = {"#00000000":"transparent", "#FFFFFFFF":"white","#F0F8FFFF":"aliceblue","#FAEBD7FF":"antiquewhite","#FFEFDBFF":"antiquewhite1","#EEDFCCFF":"antiquewhite2","#CDC0B0FF":"antiquewhite3","#8B8378FF":"antiquewhite4","#7FFFD4FF":"aquamarine","#76EEC6FF":"aquamarine2","#66CDAAFF":"aquamarine3","#458B74FF":"aquamarine4","#F0FFFFFF":"azure","#E0EEEEFF":"azure2","#C1CDCDFF":"azure3","#838 ...
+    // gets a flipped version of colsToNames so that we can search by colors for names
```
the description of the function should be describing the function, not the use for it that comes after, does that make sense?

DavidRGriswold replied:
got it. Editing. I don't have a reply box for the "test" question, but it was an accidental commit. I was testing out a command to open sublime over a network connection and created that file, then committed it one time accidentally. It shouldn't be in there now.

daattali commented on js/colourpicker.js:
```diff
@@ -668,160 +670,163 @@
-        return flipped;}
-    namesToCols = flip(colsToNames);
+      }
+      return flipped;}
```
Closing brace on new line :)

daattali commented on js/colourpicker.js:
```diff
@@ -873,87 +878,84 @@
-      rgb = correctedRGB(rgb); 
-      if (!rgb) return "";     
-      var a;
+      rgb = correctedRGB(rgb);      if (!rgb) return "";      var a;
```
3 lines

DavidRGriswold replied:
that's weird. I definitely didn't write it like that... I ran part of the code through a beautifier and now I regret that decision.

daattali commented on js/colourpicker.js:
```diff
@@ -1145,7 +1115,7 @@
       if (input.data('colourpicker-lastChange')) {
         return input.data('colourpicker-lastChange');
-      } else if (input.parent().is('.palette-limited')) {
+      } else if (input.parents('.colourpicker').is('.palette-limited')) {
```
I think `.closest()` would be a better choice, because `parents()` could potentially return multiple elements, no? We don't expect that to ever happen, but closest would make it more clear that we expect it to only match to one parent

DavidRGriswold replied:
Oooh! Yes! I didn't know about closest. My jQuery skills basically come down to what comes up when I google. =)

daattali commented on index.min.html:
```diff
@@ -77,29 +84,59 @@
     $.each($(".my-cp"), function(i, el) {
       var value = $(el).colourpicker('value');
-      $($(el).parent().parent().find("span")[0]).text(value);
+      $($(el).parents('.colourpicker').parent().find("span")[0]).text(value);
```
I made a comment about `.parents()` in the javascript file, it applies here too

daattali reviewed (commented):
Any idea what the `test` file is that got added? It shows in the commit as an empty file but I don't see it in the directory tree

daattali commented:
@DavidRGriswold I just noticed one last bug... If you type into an input box a value with an alpha, it for some reason moved the alpha value up by one.  Type `1234` and it'll give you `#11223345`. Type `#FF0000AA` and you get `#FF0000AB`. But with some values, that are very close to 0 or FF alpha, it doesn't happen

DavidRGriswold commented:
I bet this is a rounding thing.  Alpha values are rounded to the nearest
hundredth on a scale of 0 to 1, which gives less resolution than hex. I
think most tools can only render with that res, but you are right that it
is off putting. I will switch to rounding to the nearest thousandth, should
fix it.

On Jun 11, 2017 11:13 AM, "Dean Attali" <notifications@github.com> wrote:

> @DavidRGriswold <https://github.com/davidrgriswold> I just noticed one
> last bug... If you type into an input box a value with an alpha, it for
> some reason moved the alpha value up by one. Type 1234 and it'll give you
> #11223345. Type #FF0000AA and you get #FF0000AB. But with some values,
> that are very close to 0 or FF alpha, it doesn't happen
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#issuecomment-307639408>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU70cNrdf5YqOFdZoZj21ucvyLOqLZks5sDBI_gaJpZM4N1sCm>
> .
>

Commit:
Remove jQuery width code and put inputcontainer as a block element.

This puts it on the user to set the width correctly, either by setting width of colourpicker class or wrapping it in a div (which is what Shiny does!)

Commit:
commented out code correction

Commit:
minification

DavidRGriswold commented on index.html:
```diff
@@ -19,7 +19,7 @@
       margin-top: 3px;
       margin-bottom: 20px;
-      width: 30%;
+      width: 300px;
```
whoops, didn't mean to make these changes in the same commit as the minification. Arggh. OH well, I just tested the minification and it didn't change so I know I did it at the right time.

DavidRGriswold commented on js/colourpicker.js:
```diff
@@ -539,7 +539,6 @@
       input.data('colour',rgb);
       hex = rgb2hex(rgb);
       hsb = rgb2hsb(rgb);
-      var alpha = allowAlpha ? rgb.a || 1 : false;
```
this variable was unnecessary and caused a weird bug where the alpha slider didn't reset to the right place.

daattali commented:
let me know when you're done and it's ready

DavidRGriswold commented:
I am done. Even if I'm not, I'll probably need to take a break for a few
days. Gotta figure out some dart code and make some real money. =)

On Sun, Jun 11, 2017 at 9:11 PM, Dean Attali <notifications@github.com>
wrote:

> let me know when you're done and it's ready
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#issuecomment-307675013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU72b7Sm-xbliKPIgdkAJWpfbGmmG5ks5sDJ5tgaJpZM4N1sCm>
> .
>

daattali commented:
weekend is over, back to real life :(

daattali commented:
I looked at the entire code and played around with your test page a lot. One last (I'm 99% sure this is the last one!) issue: when you type in a colour into the input box, the colour palette and the hue bar get updated. But the alpha bar doesn't update.

This is what happens when I type "blue" into the box:

![image](https://user-images.githubusercontent.com/952340/27017891-02e350ec-4efa-11e7-91e7-2173958fd698.png)

Other than this, the alpha feature is ready :+1: 

(You've put in a lot of time this weekend, no rush! It can wait until you're recharged)

daattali commented:
Actually, I'll merge it and open an issue to fix that because the feature is complete as it is, this is just a bug

DavidRGriswold commented:
I'm not sure that the screenshot you gave me is a bug? blue should have the
alpha bar all the way at the top, it's alpha is 1.

I did have some versions of that bug earlier but I thought I resolved them.
Let me know.

On Sun, Jun 11, 2017 at 10:09 PM, Dean Attali <notifications@github.com>
wrote:

> Merged #1 <https://github.com/daattali/jquery-colourpicker/pull/1>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#event-1118769020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU7620wTJxttLJuTBTHGTUIStdN0hPks5sDKvggaJpZM4N1sCm>
> .
>

daattali commented:
Yes, the bar should stay at the top, but notice that the entire vertical channel didn't update to become blue.

I should have shown a better example, where you can see nothing gets updated. If you do change the alpha manually, and then type in a colour, then you'll notice the bar doesn't update.

Here's an example where I manually selected a semi transparent red, and then I typed "blue". Notice the alpha bar is still red

![image](https://user-images.githubusercontent.com/952340/27018639-fa85206e-4eff-11e7-86bf-c2fb792c936a.png)

DavidRGriswold commented:
Got it! I'm pretty sure I know what is happening there.

On Jun 11, 2017 10:45 PM, "Dean Attali" <notifications@github.com> wrote:

> Yes, the bar should stay at the top, but notice that the entire vertical
> channel didn't update to become blue.
>
> I should have shown a better example, where you can see nothing gets
> updated. If you do change the alpha manually, and then type in a colour,
> then you'll notice the bar doesn't update.
>
> Here's an example where I manually selected a semi transparent red, and
> then I typed "blue". Notice the alpha bar is still red
>
> [image: image]
> <https://user-images.githubusercontent.com/952340/27018639-fa85206e-4eff-11e7-86bf-c2fb792c936a.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/daattali/jquery-colourpicker/pull/1#issuecomment-307683813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATpU7yyZXXFOvCJNJqrEOJt7x5-Ndymzks5sDLQ9gaJpZM4N1sCm>
> .
>

daattali commented:
Great! Let's continue discussion on this issue at #2 since this PR is already merged and closed
